component/template/src/tui.rs

```diff
@@ -89,63 +89,39 @@ impl Tui {
-      let mut tick_interval = tokio::time::interval(tick_delay);
-      let mut render_interval = tokio::time::interval(render_delay);
-      _event_tx.send(Event::Init).unwrap();
+      event_tx.send(Event::Init).expect("failed to send init event"); // likely a bug if this fails
       loop {
-        let tick_delay = tick_interval.tick();
-        let render_delay = render_interval.tick();
-        let crossterm_event = reader.next().fuse();
-        tokio::select! {
-          _ = _cancellation_token.cancelled() => {
+        let event = tokio::select! {
+          _ = cancellation_token.cancelled() => {
             break;
           }
-          maybe_event = crossterm_event => {
-            match maybe_event {
-              Some(Ok(evt)) => {
-                match evt {
-                  CrosstermEvent::Key(key) => {
-                    if key.kind == KeyEventKind::Press {
-                      _event_tx.send(Event::Key(key)).unwrap();
-                    }
-                  },
-                  CrosstermEvent::Mouse(mouse) => {
-                    _event_tx.send(Event::Mouse(mouse)).unwrap();
-                  },
-                  CrosstermEvent::Resize(x, y) => {
-                    _event_tx.send(Event::Resize(x, y)).unwrap();
-                  },
-                  CrosstermEvent::FocusLost => {
-                    _event_tx.send(Event::FocusLost).unwrap();
-                  },
-                  CrosstermEvent::FocusGained => {
-                    _event_tx.send(Event::FocusGained).unwrap();
-                  },
-                  CrosstermEvent::Paste(s) => {
-                    _event_tx.send(Event::Paste(s)).unwrap();
-                  },
-                }
-              }
-              Some(Err(_)) => {
-                _event_tx.send(Event::Error).unwrap();
-              }
-              None => {},
+          crossterm_event = event_stream.next().fuse() => match crossterm_event {
+            Some(Ok(event)) => match event {
+              CrosstermEvent::Key(key) if key.kind == KeyEventKind::Press => Event::Key(key),
+              CrosstermEvent::Mouse(mouse) => Event::Mouse(mouse),
+              CrosstermEvent::Resize(x, y) => Event::Resize(x, y),
+              CrosstermEvent::FocusLost => Event::FocusLost,
+              CrosstermEvent::FocusGained => Event::FocusGained,
+              CrosstermEvent::Paste(s) => Event::Paste(s),
+              _ => continue, // ignore other events
             }
+            Some(Err(_)) => Event::Error,
+            None => break, // the event stream has stopped and will not produce any more events
           },
-          _ = tick_delay => {
-              _event_tx.send(Event::Tick).unwrap();
-          },
-          _ = render_delay => {
-              _event_tx.send(Event::Render).unwrap();
-          },
+          _ = tick_interval.tick() => Event::Tick,
+          _ = render_interval.tick() => Event::Render,
```

Comment:
What do you think about moving these to the beginning of the `select!`?

If a user presses and holds a key, with the previous and current implementation, the `Event::Tick` may never fire. If we move it to the top, I think they'll fire first before a key press happens.

Reply:
tokio::select! chooses randomly - not based on the order. See https://docs.rs/tokio/latest/tokio/macro.select.html#fairness

That said, these would look better up top from a reading perspective

Reply:
Oops, I misread this line:

> Waits on multiple concurrent branches, returning when the **first** branch completes, cancelling the remaining branches.